---
title: "Analysis of public skin metagenome data"
author: "Sander Wuyts"
date: "July 15, 2018"
output: 
  html_document:
    toc: true
    toc_float: true
---

# Introduction

In this study we will be using publicly available datasets such as in the curatedMetagenomicsData package (shotgun metagenomics) and the MicrobeDS package (for 16S rRNA gene data) to explore the presence of Lactobacilli on the skin. We will also use this data to compare with the other niches available in these datasets. In addition we will also import or own 16S rRNA gene dataset.


```{r}
suppressMessages(library(curatedMetagenomicData))
packageVersion("curatedMetagenomicData")
library(tidyverse)
suppressMessages(library(phyloseq))
packageVersion("phyloseq")
suppressMessages(library(ggpubr))
packageVersion("ggpubr")
suppressMessages(library(ggtree))
packageVersion("ggtree")
suppressMessages(library(RColorBrewer))
packageVersion("RColorBrewer")
suppressMessages(library(scales))
packageVersion("scales")
```

# Total number of samples in all projects

Let's see how many datasets of the curatedMetagenomicsProject have skin data

```{r}
n_samples <- combined_metadata %>%
  group_by(body_site) %>%
  summarise(number_of_samples = n())

combined_metadata %>%
  group_by(body_site) %>%
  summarise(count = n()) %>%
  ggplot(aes(x = body_site, fill = body_site, y = count)) +
  geom_col() +
  geom_text(aes(label = count, y = 5200)) +
  scale_fill_brewer(palette = "Dark2") +
  coord_flip() 
```

  
# Skin

There are around 500 skin samples available!  

The easiest format to work with will be the metaphlan bugs. Let's see how many datasets we can find with this dataformat:

```{r}
curatedMetagenomicData("*metaphlan_bugs*.skin", dryrun = T)
```
 
These are three different datesets from three different studies:

* Chng et al 2016: https://www.ncbi.nlm.nih.gov/pubmed/27562258
* Oh et al 2014: https://www.ncbi.nlm.nih.gov/pubmed/25279917
* Tett et al 2016: https://www.ncbi.nlm.nih.gov/pubmed/28649415 

Let's download these:

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*.skin", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

And now merge all three of them

```{r}
mergedatasets <- mergeData(datasets)

mergedatasets
```

Now I'll convert them to phyloseq and melt them in a big dataframe

```{r}
skin <- ExpressionSet2phyloseq(mergedatasets)
skin <- psmelt(skin)
```

## Analysis

Filter dataset to contain strain data only and then filter on the genus Lactobacillus:

```{r}
skin_t_lactos <- skin %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(skin_t_lactos, "out/Lactobacillus_skin_metagenomes.csv")
```

How many samples contain lactobacilli?

```{r}
skin_t_lactos %>%
  pull(Sample) %>%
  unique() %>%
  length()
```

How many of these samples have an abundance > 1?

```{r}
skin_t_lactos %>%
  filter(Abundance > 1) %>%
  pull(Sample) %>%
  unique() %>%
  length()
```

Let's plot the top 12 species based on presence absence 

```{r, fig.with = 10, fig.height= 5}
toplot <- skin_t_lactos %>%
  group_by(Species, studyID) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))

top12 <- toplot %>%
  arrange(- total_count) %>%
  pull(Species) %>%
  unique() %>%
  .[1:12] %>%
  as.character()

toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = studyID)) +
  geom_col() +
  geom_text(aes(label = total_count, y = - 5)) +
  coord_flip() + 
  scale_fill_brewer(palette = "Dark2") +
  xlab("") +
  ggtitle("Total number of Lactobacilli in skin samples") +
  theme_minimal()
```

And now let's explore their relative abundance

```{r fig.width=10, fig.height=5}
skin_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  filter(Species %in% top12) %>%
  ggplot() +
  geom_jitter(aes(x = gender, y = sumabundance, colour = Species, shape = gender), width = 0.3, size = 3) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative Abundance of Lactobacilli on Skin")
```

Is tehre a statistic significant difference?

```{r}
stats <- skin_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  select(Sample, sumabundance, gender) %>%
  distinct %>%
  spread(key = gender, value = sumabundance) 

wilcox.test(stats$male, stats$female)
```

Nope!

But we'd like to know the difference in prevalence between man and women

```{r}
stat1 <- skin %>%
  select(Sample, gender) %>%
  distinct() %>%
  group_by(gender) %>%
  summarise(total = n())
stat1
```

And finally, let's calculate how many samples of all skin samples have lactobacilli in them. We will be using this later to construct Fig1A.

```{r}
stat_metagenome <- skin %>%
  mutate(total_lactos = case_when(Sample %in% (skin %>% 
                                                 filter(Abundance > 0) %>%
                                                 filter(Genus == "Lactobacillus") %>% 
                                                 pull(Sample) %>% 
                                                 unique()) ~ "percentage_lacto",
                            !Sample %in% (skin %>% 
                                                 filter(Abundance > 0) %>%
                                                 filter(Genus == "Lactobacillus") %>% 
                                                 pull(Sample) %>% 
                                                 unique()) ~ "non_lacto")) %>%
  group_by(total_lactos) %>%
  summarise(count = n()/nrow(.)) %>%
  mutate(study = "curatedMetagenomics")

stat_metagenome
```

# Oral

Let's move on to the next niche: oralcavity. There are around 700 samples from the oral cavity available.  

Again, let's use the metaphlan_bugs format.

```{r}
curatedMetagenomicData("*metaphlan_bugs*.oralcavity", dryrun = T)
```
 
Three studies to be downloaded:

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*.oralcavity", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

And again, merge these

```{r}
mergedatasets <- mergeData(datasets)

mergedatasets
```

*Important notice*: There's a small error in the package. Some of the samples have an extremely high abundance in Lactobacillus species (up to 90%) which was suspicous to us. After contacting the authors, they confirmed me that some metadata was mixed up and that these samples were actually vaginal or nasal samples instead of oral samples. They suggested the following manual fix but promised to push a fix in the future.

```{r}
oralcavity <- ExpressionSet2phyloseq(mergedatasets)

# Need to be added to the other body sites later
extra_vaginal_samples <- subset_samples(oralcavity, body_subsite == "posterior_fornix")
extra_vaginal_samples <- psmelt(extra_vaginal_samples) %>%
  mutate(body_site = "vagina")
extra_nasal_samples <- subset_samples(oralcavity, body_subsite == "anterior_nares")
extra_nasal_samples <- psmelt(extra_nasal_samples) %>%
  mutate(body_site = "nasalcavity")

# Exclude them from oral body site
oralcavity <- subset_samples(oralcavity, body_subsite != "anterior_nares")
oralcavity <- subset_samples(oralcavity, body_subsite != "posterior_fornix")
oralcavity <- psmelt(oralcavity)
```

## Analysis

That being fixed we can now start the oral cavity analysis!

```{r}
# filter dataset for strain data only
oralcavity_t_lactos <- oralcavity %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(oralcavity_t_lactos, "out/Lactobacillus_oralcavity_metagenomes.csv")
```


What Lactobacillus species can be found in the oralcavity?

```{r, fig.with = 10, fig.height= 5}
toplot <- oralcavity_t_lactos %>%
  group_by(Species, studyID) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))

top12 <- toplot %>%
  arrange(- total_count) %>%
  pull(Species) %>%
  unique() %>%
  .[1:12] %>%
  as.character()

toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = studyID)) +
  geom_col() +
  geom_text(aes(label = total_count, y = - 5)) +
  coord_flip() + 
  scale_fill_brewer(palette = "Dark2") +
  xlab("") +
  ggtitle("Total number of Lactobacilli in oralcavity samples") +
  theme_minimal()
```

Wow, okay. The count is way lower in this niche compared to the skin. Let's look at their abundance


```{r fig.width=10, fig.height=5}
oralcavity_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  filter(Species %in% top12) %>%
  ggplot() +
  geom_jitter(aes(x = gender, y = sumabundance, colour = Species, shape = gender), width = 0.3, size = 3) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative Abundance of Lactobacilli in oralcavity")
```

Very low abundance

# Nasal

On to the next niche, the nasal niche!

Again the same story:

```{r}
curatedMetagenomicData("*metaphlan_bugs*.nasalcavity", dryrun = T)
```
 
Let's download this dataset:

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*.nasalcavity", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

And merge it

```{r}
mergedatasets <- mergeData(datasets)

mergedatasets
```

And finally convert to phyloseq. *Important remark*: We'll be adding the samples we removed from the oral dataset here to the nasal!

```{r warning=FALSE}
nasalcavity <- ExpressionSet2phyloseq(mergedatasets)
nasalcavity <- psmelt(nasalcavity) %>%
  bind_rows(extra_nasal_samples)
```

## Analysis

Can we find lactobacilli in the nose?


```{r}
# filter dataset for strain data only
nasalcavity_t_lactos <- nasalcavity %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(nasalcavity_t_lactos, "out/Lactobacillus_nasalcavity_metagenomes.csv")
```

Top 12 plot

```{r, fig.with = 10, fig.height= 5}
toplot <- nasalcavity_t_lactos %>%
  group_by(Species, studyID) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))

top12 <- toplot %>%
  arrange(- total_count) %>%
  pull(Species) %>%
  unique() %>%
  .[1:12] %>%
  as.character()

toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = studyID)) +
  geom_col() +
  geom_text(aes(label = total_count, y = - 5)) +
  coord_flip() + 
  scale_fill_brewer(palette = "Dark2") +
  xlab("") +
  ggtitle("Total number of Lactobacilli in nasal samples") +
  theme_minimal()
```

That's not a lot...

```{r fig.width=10, fig.height=5}
nasalcavity_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  filter(Species %in% top12) %>%
  ggplot() +
  geom_jitter(aes(x = gender, y = sumabundance, colour = Species, shape = gender), width = 0.3, size = 3) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative Abundance of Lactobacilli in nasal cavity")
```

Some of them have a remarkable high relative abundance.  

However, due to the very low sample numbers we will not be using them.

# Milk

What about mother's milk?

```{r}
curatedMetagenomicData("*metaphlan_bugs*.milk", dryrun = T)
```
 

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*.milk", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

```{r}
mergedatasets <- mergeData(datasets)

mergedatasets
```

```{r}
milk <- ExpressionSet2phyloseq(mergedatasets)
milk <- psmelt(milk)
```

## Analysis

Can we find lactobacilli here?


```{r}
# filter dataset for strain data only
milk_t_lactos <- milk %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(milk_t_lactos, "out/Lactobacillus_milk_metagenomes.csv")
```

Nope, the dataset is completely empty...


# Vaginal

And then now, the vaginal niche! Same story as above.

```{r}
curatedMetagenomicData("*metaphlan_bugs*vagina", dryrun = T)
```
 
Downloading

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*vagina", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

Merging of datasets 

```{r}
mergedatasets <- mergeData(datasets)

mergedatasets
```

Let's convert to phyloseq. *Important note*: we will be adding the samples that were wrongly annotated as belonging to the oral cavity here!

```{r warning=FALSE}
vagina <- ExpressionSet2phyloseq(mergedatasets)
vagina <- psmelt(vagina) %>%
  bind_rows(extra_vaginal_samples)
```

## Analysis

How many lactobacillus species can we find?


```{r}
# filter dataset for strain data only
vagina_t_lactos <- vagina %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(vagina_t_lactos, "out/Lactobacillus_vagina_metagenomes.csv")
```

Top12 species plot:

```{r, fig.with = 10, fig.height= 5}
toplot <- vagina_t_lactos %>%
  group_by(Species, studyID) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))

top12 <- toplot %>%
  arrange(- total_count) %>%
  pull(Species) %>%
  unique() %>%
  .[1:12] %>%
  as.character()

toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = studyID)) +
  geom_col() +
  geom_text(aes(label = total_count, y = - 5)) +
  coord_flip() + 
  scale_fill_brewer(palette = "Dark2") +
  xlab("") +
  ggtitle("Total number of Lactobacilli in vaginal samples") +
  theme_minimal()
```

Relative abundance

```{r fig.width=10, fig.height=5}
vagina_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  filter(Species %in% top12) %>%
  ggplot() +
  geom_jitter(aes(x = gender, y = sumabundance, colour = Species, shape = gender), width = 0.3, size = 3) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative Abundance of Lactobacilli in vagina")
```

As expected, this is a lot higher!

# Stool

The final niche will be the human stool niche. Sorry for the repetition, I know it's probably getting boring.

```{r}
curatedMetagenomicData("*metaphlan_bugs*stool", dryrun = T)
```
 
Damn! That's way more data than all other datasets! Here I will be downloading them:

```{r}
datasets <- curatedMetagenomicData("*metaphlan_bugs*stool", dryrun = F)

for (i in 1:length(datasets)){
  print(str_c("Dataset ",i))
  print(datasets[i])
  print(datasets[[i]])
}
```

As the size of these datasets is much higher than the others, I will loop over 2 datasets at one time and filter them for Lactobacilli. Only the samples that have Lactobacillus in them will be kept. In the loop I will also do the conversion to phyloseq and the large dataframe.

```{r eval=FALSE}
stool <- tibble()
for (i in 1:(length(datasets)-1)){
  stool_temp <- mergeData(datasets[i:(i+1)]) %>%
    ExpressionSet2phyloseq() %>%
    psmelt() %>%filter(str_detect(OTU, "t__"))  %>%
    filter(Genus == "Lactobacillus") %>%
    filter(Abundance > 0) %>%
    replace_na(list(disease = "Unkown"))
  
  stool <- stool %>% 
    bind_rows(stool_temp)
}

saveRDS(stool, 'in/stool.rds')
```

To avoid doing this hard task everytime again, I will just read in a previously saved rds file.

```{r}
stool <- readRDS("in/stool.rds")
```

## Analysis

Ready for analysis!


```{r}
# filter dataset for strain data only
stool_t_lactos <- stool %>%
  filter(str_detect(OTU, "t__"))  %>%
  filter(Genus == "Lactobacillus") %>%
  filter(Abundance > 0) %>%
  replace_na(list(disease = "Unkown"))

write.csv(stool_t_lactos, "out/Lactobacillus_stool_metagenomes.csv")
```

Top 12 species

```{r, fig.with = 10, fig.height= 10}
toplot <- stool_t_lactos %>%
  group_by(Species, studyID) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))

top12 <- toplot %>%
  arrange(- total_count) %>%
  pull(Species) %>%
  unique() %>%
  .[1:12] %>%
  as.character()

toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = studyID)) +
  geom_col() +
  geom_text(aes(label = total_count, y = - 5)) +
  coord_flip() + 
  xlab("") +
  ggtitle("Total number of Lactobacilli in stool samples") +
  theme_minimal() +
  theme(legend.position = "bottom")
```

Wooooow! This is something completely different!

```{r fig.width=10, fig.height=5}
stool_t_lactos %>%
  group_by(Sample) %>%
  mutate(sumabundance = sum(Abundance)) %>%
  filter(Species %in% top12) %>%
  ggplot() +
  geom_jitter(aes(x = gender, y = Abundance, colour = Species, shape = gender), width = 0.3, size = 3) +
  scale_color_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative Abundance of Lactobacilli in stool")
```


# Combine all niches

Now that we went through all niches, let's combine these into one big file and plot!

```{r fig.height=10, fig.width=12, warning=FALSE}
toplot <- oralcavity_t_lactos %>%
  bind_rows(nasalcavity_t_lactos)  %>%
  bind_rows(skin_t_lactos) %>%
  bind_rows(stool_t_lactos) %>%
  bind_rows(vagina_t_lactos) %>%
  group_by(Species, body_site) %>%
  summarise(count = n()) %>%
  group_by(Species) %>%
  mutate(total_count = sum(count))


toplot %>%
  ggplot(aes(x = reorder(Species, - total_count), y = count, fill = body_site)) +
  geom_col() +
  coord_flip() + 
  scale_fill_brewer(palette = "Dark2") +
  xlab("") +
  ggtitle("Total number of Lactobacilli per body site") +
  theme_minimal() +
  facet_grid(~body_site, scales = 'free_x', space = "free_y")
```

Cool stuff! Let's do the same for abundance

```{r warning=FALSE}
oralcavity_t_lactos %>%
  bind_rows(nasalcavity_t_lactos)  %>%
  bind_rows(skin_t_lactos) %>%
  bind_rows(stool_t_lactos) %>%
  bind_rows(vagina_t_lactos) %>% 
#  filter(Abundance > 1) %>%
  ggplot(aes(x = body_site, y = Abundance, fill = body_site)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(colour = 'black', shape = 21, size = 0.7, alpha = 0.5, width = 0.2) +
  scale_fill_brewer(palette = "Paired") +
  theme_bw() +
  ggtitle("Relative abundance of Lactobacilli in all body sites")
```

# Own 16S dataset

We'll also be using our own dataset, namely skin samples collected from cheeck swabs. For this I will read in a csv that my colleague prepared.

```{r}
partial_16S <- read_csv2("in/baselineCon_ASVlevel.csv", col_types = cols(gender = col_character())) 
```

How many samples do we have in this dataset?

```{r}
stat1 <- partial_16S %>%
  select(sample, gender) %>%
  distinct %>%
  group_by(gender) %>%
  summarise(total = n())
stat1
```

How many of these sample contain lactobacilli? Let's save this for plotting later in Figure 1A

```{r}
stat_16S <- partial_16S %>%
  mutate(total_lactos = case_when(sample %in% (partial_16S %>% 
                                                 filter(rel_abundance > 0) %>%
                                                 filter(genus == "Lactobacillus") %>% 
                                                 pull(sample) %>% 
                                                 unique()) ~ "percentage_lacto",
                            !sample %in% (partial_16S %>% 
                                            filter(rel_abundance > 0) %>%
                                            filter(genus == "Lactobacillus") %>% 
                                            pull(sample) %>% 
                                            unique()) ~ "non_lacto")) %>%
  group_by(total_lactos) %>%
  summarise(count = n()/nrow(.)) %>%
  mutate(study = "This study")
stat_16S 
```

High percentage of lactos!

# 16S rRNA dataset from HMP

Now we will add the V35 dataset from the Human Microbiome Project (HMP) with this usefull package: MicrobeDs (https://github.com/twbattaglia/MicrobeDS). They made the dataset available Phyloseq ready!

```{r message=FALSE, warning=FALSE}
# Load Library
library(MicrobeDS)

# Load selected datasets
data('HMPv35')

# Calculate relative abundance
HMPv35_relabun <- transform_sample_counts(HMPv35, function(x) x / sum(x))

# Check number of samples
nsamples(HMPv35_relabun)
```

That's a nice addition to our analysis!


Next, we'll be filtering to keep only Lactobacillus OTUs.

```{r message=FALSE, warning=FALSE}
# Subset on lactos and use this for visualisation
HMPv35_lactos <- subset_taxa(HMPv35_relabun, Genus == "Lactobacillus") %>% 
  psmelt() %>%
  filter(Abundance > 0)
```

Now let's subset the original dataset on Skin as well and then annotate calculate how many samples contain lactos.

```{r message=FALSE}
# Subset on skin
HMPv35_skin <- subset_samples(HMPv35_relabun, env == "Skin")
HMPv35_skin_df <- data.frame(sample_data(HMPv35_skin)) %>%
      rownames_to_column(var = "Sample")

HMP_stats <- HMPv35_skin_df %>%
  mutate(total_lactos = case_when(Sample %in% (HMPv35_lactos %>% pull(Sample) %>% unique()) ~ "percentage_lacto",
                            !Sample %in% (HMPv35_lactos %>% pull(Sample) %>% unique()) ~ "non_lacto")) %>%
  group_by(total_lactos) %>%
  summarise(count = n()/nrow(.)) %>%
  mutate(study = "HMPv35")
```

# Fig 1A

Now we've got all necessary data to start constructing our figure. Let's start with Figure 1A.

```{r}
Fig1A <- stat_16S %>% 
  bind_rows(stat_metagenome) %>%
  bind_rows(HMP_stats) %>%
  mutate(total_lactos = case_when(total_lactos == "non_lacto" ~ "No Lactobacilli detected",
                            total_lactos == "percentage_lacto" ~ "Lactobacilli detected")) %>%
  mutate(total_lactos = factor(total_lactos, levels = c("No Lactobacilli detected",
                                                           "Lactobacilli detected"))) %>%
  mutate(study = factor(study, levels = c("This study", "HMPv35", "curatedMetagenomics"))) %>%
  mutate(count = count*100) %>%
  ggplot(aes(x = study, y = count, fill = total_lactos)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(breaks = c(0, 50, 100)) +
  scale_fill_brewer(palette = "Paired") +
  ylab("\n Percentage of samples (%)") +
  theme_minimal() +
  theme(strip.text.x = element_text(face = "bold", size = 14),
        
        axis.text.x = element_text(size = 8, angle = 45, vjust = 1, hjust = 1, face = "bold"),
        axis.text.y = element_text(face = "bold", size = 10),
        
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 6),
        
        axis.ticks = element_line(),
        
        legend.title = element_blank(),
        
        panel.grid = element_blank(),
        
        legend.direction = "vertical",
        legend.position = "right", 
        legend.justification = "left",
        legend.margin = margin(0,0,0,0),
        legend.text = element_text(size = 6),
        legend.key.size = unit(6, "pt")) +
  ggtitle("")

Fig1A  
```

This figure shows the percentage of skin microbiome samples that have lactobacilli in them from the three different studies.

## Fig1B

In Fig1B we will be looking at the relative abundances. For this graph we will be summing all Lactobacillus relative abundances per sample. We thus get one value per sample per body niche.  

First we'll do this for the HMP 16S dataset. 


```{R}
HMP_plot <- HMPv35_lactos %>%
  distinct() %>%
  group_by(Sample, env) %>%
  summarise(summedAbundance = sum(Abundance)*100) %>%
  ungroup() %>%
  add_count(env) %>%
  mutate(study = "HMPv35") %>%
  rename(body_site = env) %>%
  mutate(body_site = as.character(body_site)) %>%
  filter(summedAbundance > 0)
```

Now also for the above described datasets from the curatedMetagenomics package, except for the nasal cavity as not a lot of lacto samples were detected there.

```{r}
curatedMetagenomics <- oralcavity_t_lactos %>%
#  bind_rows(nasalcavity_t_lactos)  %>% # Nasal cavity is filtered out because only 3 samples contain lactos
  bind_rows(skin_t_lactos) %>%
  bind_rows(stool_t_lactos) %>%
  bind_rows(vagina_t_lactos) %>% 
  filter(Abundance > 0) %>%
  distinct() %>%
  group_by(Sample, body_site) %>%
  summarise(summedAbundance = sum(Abundance)) %>%
  ungroup() %>%
  add_count(body_site) %>%
  mutate(study = "curatedMetagenomics") %>%
  mutate(body_site = case_when(body_site == "stool" ~ "Stool",
                               body_site == "skin" ~ "Skin",
                               body_site == "vagina" ~ "Vaginal",
#                               body_site == "nasalcavity" ~ "Nasal",
                               body_site == "oralcavity" ~ "Oral"))
```

And finally also for our own dataset.

```{R}
b_16S <- partial_16S %>%
  filter(genus == "Lactobacillus") %>%
  group_by(sample, gender) %>%
  summarise(summedAbundance = sum(rel_abundance)*100) %>%
  mutate(study = "This study") %>%
  mutate(body_site = "Skin") %>%
  mutate(n = nrow(.))
```

Before we start constructing our Fig 1B we also need to do some counting for the labels. Then we'll go on with constructing Figure 1B

```{R}
fig1B_count <- curatedMetagenomics %>%
  bind_rows(HMP_plot) %>%
  bind_rows(b_16S) %>%
  select(body_site, n, study) %>%
  distinct() %>%
  mutate(study = factor(study, levels = c("This study", "HMPv35", "curatedMetagenomics")))
 

label_format <- format_format(big.mark = " ", decimal.mark = ".", scientific = FALSE)

Fig1B <- curatedMetagenomics %>%
  bind_rows(HMP_plot) %>%
  bind_rows(b_16S) %>%
  mutate(study = factor(study, levels = c("This study", "HMPv35", "curatedMetagenomics"))) %>%
  mutate(body_site = factor(body_site, levels = c("Skin", "Oral", "Stool", "Vaginal"))) %>%
  ggplot(aes(x = body_site, y = summedAbundance, fill = study)) +
  geom_boxplot(outlier.alpha = 0, alpha = 0.4) +
  geom_point(aes(colour = study), position = position_jitterdodge(jitter.height = 0, jitter.width = 0.1),
             shape = 21, size = 0.7, alpha = 0.2) +
  geom_text(data = fig1B_count, aes(y = 250, label = str_c("n = ", n), group = study), position = position_dodge(width = 0.8), size = 2, angle = 45, hjust= 0, vjust= 0) +
  scale_fill_brewer(palette = "Dark2") +
  scale_colour_brewer(palette = "Dark2") +
  expand_limits(y = 8000) +
  scale_y_log10(labels = label_format, breaks = c(0.01,1,100)) +
  ylab("\n Relative abundance (%)") +
  theme_minimal() +
  theme(strip.text.x = element_text(face = "bold", size = 14),
        
        axis.text.x = element_text(size = 8, angle = 45, hjust = 1, vjust =1, face = "bold"),
        axis.text.y = element_text(face = "bold", size = 12),
        
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 8),
        
        axis.ticks = element_line(),
        
        legend.title = element_blank(),
        legend.position = 'right',
        
        panel.grid.major.y =   element_line(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank()
        
        ) +
  ggtitle("") 

Fig1B
```

# Fig 1E

We need some information from 1E to colour figure 1C. Therefore we will be plotting 1E first.  

In this figure we will be counting the top 12 Lactobacillus species from the shotgun metagenomic dataset. 

```{r}
curatedMetagenomics_top12_species <- skin_t_lactos %>%
  group_by(Species) %>%
  summarise(count = n()) %>%
  separate(Species, into = c('genus', 'species'), sep ="_", extra = "merge") %>% 
  unite(Species, c("genus", "species"), sep = " ") %>%
  arrange(-count) %>%
  .[1:12,]
  
# Create table that connects species to the phylogenetic group they belong to
species_to_phylogenetic_group <- tibble(Species = c("Lactobacillus crispatus",
                                                    "Lactobacillus iners",
                                                    "Lactobacillus gasseri",
                                                    "Lactobacillus jensenii",
                                                    "Lactobacillus plantarum",
                                                    "Lactobacillus delbrueckii",
                                                    "Lactobacillus fermentum",
                                                    "Lactobacillus casei_paracasei",
                                                    "Lactobacillus sakei",
                                                    "Lactobacillus vaginalis",
                                                    "Lactobacillus rhamnosus",
                                                    "Lactobacillus curvatus"),
                                        clade_name = c("L. delbrueckii group",
                                                       "L. delbrueckii group",
                                                       "L. delbrueckii group",
                                                       "L. delbrueckii group",
                                                       "L. plantarum group",
                                                       "L. delbrueckii group",
                                                       "L. reuteri group",
                                                       "L. casei group",
                                                       "L. sakei group",
                                                       "L. reuteri group",
                                                       "L. casei group",
                                                       "L. sakei group"))


# Add this data to the dataset
curatedMetagenomics_top12_species <- curatedMetagenomics_top12_species %>%
  left_join(species_to_phylogenetic_group) 


Fig1E <- curatedMetagenomics_top12_species %>%
  ggplot(aes(x = reorder(Species, count), y = count, fill = clade_name)) +
  geom_col() +
  coord_flip() +
  ylab("Count") +
  xlab("") + 
  theme(
    # Set text size
    axis.title.y = element_text(size = 1,
                                margin = margin(0,0,0,0)),
    axis.title.x = element_text(size = 10, 
                                margin = margin(5,0,0,0)),
    
    axis.text.x = element_text(size = 6, 
                               colour="black",
                               margin = margin(2,0,0,0)),
    axis.text.y = element_text(size = 6,
                               colour="black",
                               margin = margin(0,0,0,0),
                               face = "italic"),

    # Configure lines and axes
    axis.ticks.x = element_line(colour = "black"), 
    axis.ticks.y = element_blank(), 
    
    # Plot background
    panel.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(), 
    
    # Remove border
    panel.border = element_rect(fill="transparent",color=NA),
    
    # Remove legend
    legend.position =  "right",
    legend.title = element_blank()
    )

Fig1E
```

We'll fix the colours later

# Fig 1C

Time to do the phylogenetic placement. This is done in a separate script, but here we will be writing the ASVs to a file that we will use as input for the placement.

```{r}
partial_16S %>% 
  filter(genus == "Lactobacillus") %>%
  select(taxon, family, genus, taxon_name) %>%
  distinct() %>%
  write_tsv("phylogenetic_placement/input/ASVs.tsv")
```

Now time for the phylogenetic placement! This is probably the hardest part to understand from this whole analysis as we need to do a lot of parsing. But basically we will be reading in a jplace object (which contains oru phylogenetic placements and our 16S rRNA tree) and convert this to a large dataframe which we called tidytree. This makes data handling and mutating much easier. We then manipulate this object to figure out to be able to annotate clades and colour the branches that have a sequenced placed on them.

```{r}
# functions by s. wittouck to convert a jplace object to a tidytree object
jplace2tidytrees <- function(jplace, outgroup_tips) {
  
  jplace@phylo$edge.length <- tibble(node = jplace@phylo$edge[, 2]) %>%
    left_join(attr(jplace@phylo, "edgeNum")) %>%
    pull(edgeNum)
  
  outgroup_node <- MRCA(jplace@phylo, tip = outgroup_tips)
  
  jplace@phylo <- reroot(jplace@phylo, outgroup_node)
  
  nodes <- tibble()
  
  for (taxon_index in 1:length(jplace@placements$n)) {
    
    taxon_name <- jplace@placements$n[[taxon_index]]
    
    nodes_new <- as.data.frame(jplace@phylo, branch.length = "none") %>%
      rename(edge_num = length)
    
    placements <- jplace@placements$p[[taxon_index]] %>%
      as_tibble() %>%
      set_names(jplace@fields)
    
    nodes_new <- left_join(nodes_new, placements) %>%
      mutate(placement = ifelse(is.na(like_weight_ratio), "no", "yes")) %>%
      mutate(asv = !! taxon_name)
    
    nodes <- bind_rows(nodes, nodes_new)
    
  }
  
  list(tidytrees = nodes, tree = jplace@phylo)
  
}

# Read in the phylogenetic placement data
jplace <- read.jplace(file = "phylogenetic_placement/data/tree_placement/RAxML_portableTree.placement.jplace")

outgroup_tips <- c(
  "GCA_001437585.1_1",
  "GCA_001438885.1_1",
  "GCA_001437015.1_1",
  "GCA_001437015.1_2"
)

out <- jplace2tidytrees(jplace, outgroup_tips)
tidytrees <- out$tidytrees
tree <- out$tree

# Read in species name for all used genomes
genomes <- read_tsv(file = "phylogenetic_placement/data/genomes.tsv") 


# Figure out the best nodes per placed ASV.
asvs_best_node <- tidytrees %>%
  group_by(asv) %>%
  filter(like_weight_ratio == max(like_weight_ratio, na.rm = T)) %>%
  slice(1) %>%
  ungroup() %>%
  select(asv, best_node = node)

# Optimise tidytree
tidytree <- tidytrees %>%
  group_by(node, parent, edge_num, x, y, label, isTip, branch, angle) %>%
  summarize(placement = sum(! is.na(like_weight_ratio)) > 0) %>%
  mutate(placement = ifelse(placement, "yes", "no")) %>%
  mutate(best_placement = ifelse(node %in% asvs_best_node$best_node, "yes", "no")) %>%
  mutate(assembly_accession = str_extract(label, "[A-Z]+_[0-9]+\\.[0-9]")) %>%
  left_join(genomes) %>%
  mutate(species = str_replace_all(species, "_", " ")) %>%
  mutate(species = str_extract(species, "[^ ]+ [^ ]+"))

# find the node numbers of clade MRCAs
tidytree  %>%
  ggtree(layout = "circular", aes(col = placement), size = 0.5, branch.length = "none") +
  geom_tiplab2(aes(label = node), col = "#000000", size = 3) +
  scale_color_manual(values = c(no = "#D0D0D0", yes = "#d95f02")) +
  theme(plot.title = element_text(hjust = 0.5)) 

# Define clades based on these numbers
clades <- tibble(
    tip1 = c(1, 120, 153, 166, 177),
    tip2 = c(23, 115, 165, 170, 177),
    clade_name = c(
      "L. delbrueckii group",
      "L. plantarum group",
      "L. casei group",
      "L. sakei group",
      "L. algidus group"
    )
  ) %>%
  mutate(mrca = map2_int(tip1, tip2, function(x, y) MRCA(tree, c(x, y)))) %>%
  mutate(col = brewer.pal(nrow(.), name = "Paired") %>% as.character())


nodes_clades <- clades %>%
  group_by(mrca, clade_name) %>%
  do({
    tibble(node = c(getDescendants(tree = tree, node = .$mrca), .$mrca))
  }) %>%
  ungroup() %>%
  rename(clade = mrca)

colors <- clades$col %>%
  set_names(clades$clade_name %>% as.character()) %>%
  c(., "other" = "#D0D0D0") %>%
  c(., "unclassified" = "#000000") %>%
  c(., "L. reuteri group" = "#E31A1C")

# This is necessary for colouring the tips according to the curateddMetagenomics dataset
curatedMetagenomics_top12_species <- curatedMetagenomics_top12_species %>%
  mutate(Species = if_else(Species == "Lactobacillus casei_paracasei", "Lactobacillus casei", Species)) %>%
  rename(tipcolor = clade_name) %>%
  rename(species = Species)


# Plot Fig 1C
Fig1C <- tidytree  %>%
  left_join(nodes_clades) %>%
  left_join(curatedMetagenomics_top12_species) %>%
  mutate(clade_name = ifelse(is.na(clade_name), "unclassified", clade_name)) %>%
  mutate(tipcolor = ifelse(is.na(tipcolor), "unclassified", tipcolor)) %>%
  mutate(color = ifelse(placement == "yes", clade_name, "other")) %>%
  ggtree(layout = "circular", aes(col = color), size = 1, branch.length = "none") +
  geom_tiplab2(aes(label = species, colour = tipcolor), size = 1.2, offset = 0.8) +
  scale_color_manual(values = colors,
                     breaks = c("L. delbrueckii group",
                                "L. plantarum group",
                                "L. casei group",
                                "L. sakei group",
                                "L. algidus group",
                                "unclassified"),
                     name = "",
                     labels = c(expression(paste(italic("Lactobacillus delbrueckii")," group")),
                                expression(paste(italic("Lactobacillus plantarum")," group")),
                                expression(paste(italic("Lactobacillus casei")," group")),
                                expression(paste(italic("Lactobacillus sakei")," group")),
                                expression(paste(italic("Lactobacillus algidus")," group")),
                                "Unclassified")) +
  theme(
    legend.position = "none",
    legend.text.align = 0
  ) +
  ggplot2::xlim(0,28)

Fig1C
```
# Fig 1 D

And then as a final figure we'll be constructing 1D which shows how many times a sequence was placed in a certain clade.

```{R}
# Figure out to which group the best node per ASV belongs

Fig1D <- asvs_best_node %>%
  left_join(nodes_clades, by = c("best_node" =  "node")) %>%
  group_by(clade_name) %>%
  add_count() %>%
  select(clade_name, n) %>%
  distinct() %>%
  ungroup() %>%
  ggplot(aes(x = reorder(clade_name, n), y = n, fill = clade_name)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  scale_fill_manual(values = colors) +
#  scale_y_continuous(breaks = c(0,2,4,6,8)) +
  scale_x_discrete(breaks = c("L. delbrueckii group",
                                "L. plantarum group",
                                "L. casei group",
                                "L. sakei group",
                                "L. algidus group"
                                ),
                     labels = c(expression(paste(italic("Lactobacillus delbrueckii")," group")),
                                expression(paste(italic("Lactobacillus plantarum/pentosus")," group")),
                                expression(paste(italic("Lactobacillus (para)casei/rhamnosus")," group")),
                                expression(paste(italic("Lactobacillus sakei")," group")),
                                expression(paste(italic("Lactobacillus algidus")," group"))))+
  ylab("Count") +
  xlab("") + 
  theme(
    # Set text size
    axis.title.y = element_text(size = 1,
                                margin = margin(0,0,0,0)),
    axis.title.x = element_text(size = 10, 
                                margin = margin(5,0,0,0)),
    
    axis.text.x = element_text(size = 6, 
                               colour="black",
                               margin = margin(2,0,0,0)),
    axis.text.y = element_text(size = 6,
                               colour="black",
                               margin = margin(0,0,0,0),
                               face = "italic"),

    # Configure lines and axes
    axis.ticks.x = element_line(colour = "black"), 
    axis.ticks.y = element_blank(), 
    
    # Plot background
    panel.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(), 
    
    # Remove border
    panel.border = element_rect(fill="transparent",color=NA),
    
    # Remove legend
    legend.position =  "none",
    
    # Align title
    plot.title = element_text(hjust = 2.5)
    ) +
  ggtitle("This study")

Fig1D
```

# Fig1E_recoloured

Now that we have these colours we can recolour figure 1E.

```{r}
Fig1E <- Fig1E + 
  geom_blank(data = tibble(Species = "Lactobacillus crispatus", clade_name = "L. algidus group", count = 1)) + 
  scale_fill_manual(values = colors,
                    labels = c(expression(paste(italic("L. algidus")," group")),
                               expression(paste(italic("L. (para)casei/rhamnosus")," group")),
                               expression(paste(italic("L. delbrueckii")," group")),
                               expression(paste(italic("L. plantarum/pentosus")," group")),
                               expression(paste(italic("L. reuteri")," group")),
                               expression(paste(italic("L. sakei")," group"))
                               )) +
  ggtitle("CuratedMetagenomics") +
  theme(legend.text.align = 0,
        legend.text = element_text(size = 6),
        legend.key.size = unit(6, "pt"))

  

Fig1E
```


# Figure 1

Finally the: the grand finale!

```{r}
ggarrange(ggarrange(Fig1A,
                    Fig1B,
                    ncol = 2,
                    widths = c(1,2),
                    labels = c("a", "b")),
          ggarrange(Fig1C,
                    ncol = 1,
                    labels = c("c")),
          
          ggarrange(Fig1D,
                    Fig1E,
                    labels = c("d", "e"),
                    ncol = 2,
                    widths = c(1,2)
                    ),
          heights = c(1,2,1),
          nrow = 3)

# Low quality
ggsave(filename = "Figure1.png",
       width = 178,
       height = 210,
       units = "mm",
       dpi = 300)

# Print quality
# ggsave(filename = "Figure1.tiff",
#        width = 178,
#        height = 210,
#        units = "mm",
#        dpi = 300)
```


